components/script/dom/textcontrol.rs

```diff
@@ -3,18 +3,33 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use dom::bindings::cell::DomRefCell;
+use dom::bindings::codegen::Bindings::HTMLFormElementBinding::SelectionMode;
 use dom::bindings::conversions::DerivedFrom;
 use dom::bindings::error::{Error, ErrorResult};
 use dom::bindings::str::DOMString;
 use dom::event::{EventBubbles, EventCancelable};
 use dom::eventtarget::EventTarget;
 use dom::node::{Node, NodeDamage, window_from_node};
 use script_traits::ScriptToConstellationChan;
-use textinput::{SelectionDirection, TextInput};
+use std::cell::Cell;
+use textinput::{SelectionDirection, SelectionState, TextInput};
 
 pub trait TextControl: DerivedFrom<EventTarget> + DerivedFrom<Node> {
     fn textinput(&self) -> &DomRefCell<TextInput<ScriptToConstellationChan>>;
     fn selection_api_applies(&self) -> bool;
+    fn has_selectable_text(&self) -> bool;
+    fn dirty_value_flag(&self) -> &Cell<bool>;
```

Comment:
Returning a &Cell<bool> here looks like an implementation detail, rather than a generalized API. I would expect this to return a bool, and have another method that sets the dirty value flag.